abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+  - `i32 (const char*) message_data`
+  - `i32 (size_t) message_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Logs message (`message_data`, `message_size`) at the log level (`log_level`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `log_level`.
+
+Returns `InvalidMemoryAccess` when `message_data` and/or `message_size` point to
+invalid memory address.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_memory_allocate`
+
+* params:
+  - `i32 (size_t) memory_size`
+* returns:
+  - `i32 (void*) memory_data`
+
+Allocates continuous memory buffer of size `memory_size` using the in-VM memory
+allocator and returns address of the memory buffer to the host environment using
+`memory_data`. Returning `0` indicates failure.
+
+
+## Context lifecycle
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_context_create`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) parent_context_id`
+  - `i32 (proxy_context_type_t) context_type`
+* returns:
+  - `i32 (uint32_t) new_context_id`
+
+Called when the host environment creates a new context (`context_id`) of type
+`context_type` (e.g. `VmContext`, `PluginContext`, `StreamContext`,
+`HttpContext`).
+
+This callback can be used to associate `context_id` with `parent_context_id`,
+or when using custom context identifiers, but otherwise the in-VM context can be
+implicitly created in `proxy_on_vm_start`, `proxy_on_plugin_start`,
+`proxy_on_new_connection` and `proxy_on_http_request_headers` callbacks.
+
+Return value (`new_context_id`) can be used to assign new context identifier
+that will be used by the host environment in subsequent calls. Returning `0`
+retains the original context identifier.
+
+#### `proxy_on_context_finalize`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - `i32 (bool) is_done`
+
+Called when the host environment is done processing context (`context_id`).
+Any finalization (e.g. logging) should be done within this callback.
+
+Return value (`is_done`) indicates that the Wasm module also finished processing
+context (`context_id`). If it didn't, then it should return `false` and call
+`proxy_context_finalize` when it's done.
+
+### Functions exposed by the host environment
+
+#### `proxy_set_effective_context`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Change the effective context to `context_id`. This function is usually used to
+change effective context during `proxy_on_queue_ready`, `proxy_on_timer_ready`,
+`proxy_on_http_call_response` and `proxy_on_grpc_call_response` callbacks.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `context_id`.
+
+#### `proxy_context_finalize`
+
+* params:
+  - none
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Indicates to the host environment that the Wasm module finished processing
+current context. It should be used after previously returning `false` in
```

Comment:
nit: you might want to mention `proxy_set_effective_context` which I assume might need to be called if the contexts have switched before finalizing.